src/posix/net.cpp

```diff
@@ -1680,53 +1770,71 @@ bool CurlHttpIO::doio()
     bool result;
     statechange = false;
 
-#if !defined(_WIN32) || defined(WINDOWS_PHONE)
-    if (waiter)
-    {
-        ares_process(ares, &waiter->rfds, &waiter->wfds);
-    }
-#else
-    for (unsigned int i = 0; i < aressockets.size(); i++)
-    {
-        SockInfo &info = aressockets[i];
-        ares_process_fd(ares,
-                        info.mode & SockInfo::READ ? info.fd : ARES_SOCKET_BAD,
-                        info.mode & SockInfo::WRITE ? info.fd : ARES_SOCKET_BAD);
-    }
-#endif
+    processaresevents();
+    closearesevents();
+
     result = statechange;
     statechange = false;
 
-#if !defined(_WIN32) || defined(WINDOWS_PHONE)
-    int dummy = 0;
-    curl_multi_perform(curlmapi, &dummy);
-#else
+    if (curlsocketsprocessed)
+    {
+        return result;
+    }
+
     processcurlevents(API);
-#endif
     result |= multidoio(curlmapi);
 
-    if (!maxdownloadspeed || maxdownloadspeed > downloadSpeed)
+    partialdownloaddata = 0;
+    if (aredownloadspaused)
+    {
+        aredownloadspaused = false;
+        while (pauseddownloads.size() && !aredownloadspaused)
```

Comment:
I think that it should be safe, because insertions are controlled by `aredownloadspaused` and deletions should never happen there. I will check it.